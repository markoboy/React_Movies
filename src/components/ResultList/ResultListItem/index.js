export { default } from './ResultListItem.component';
