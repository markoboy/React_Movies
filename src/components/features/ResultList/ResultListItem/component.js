/* eslint-disable camelcase */
/* eslint-disable jsx-a11y/anchor-is-valid */
import Image from '@components/common/Image';
import { MOVIE_FALLBACK_IMAGE, MOVIE_POSTER_HEIGHT, MOVIE_POSTER_WIDTH } from '@constants/Generic';
import { SharedMovieType } from '@constants/MovieTypes';
import React from 'react';
import { Link } from 'react-router-dom';
import './styles.scss';

function ResultListItemComponent({
  id,
  poster_path,
  title,
  release_date,
  genres,
  children,
}) {
  return (
    <li className="result-list__item" data-testid="result-list-item">
      <Link className="result-item-container" to={`/film/${id}`}>
        <Image
          src={poster_path}
          width={MOVIE_POSTER_WIDTH}
          height={MOVIE_POSTER_HEIGHT}
          alt={title}
          fallbackSrc={MOVIE_FALLBACK_IMAGE}
        />

        <div className="result-item__body flex flex--wrap">
          <h4 className="result-item__title">{title}</h4>
          <p className="result-item__release-date">
            {release_date?.getFullYear()}
          </p>
          <p className="result-item__genre">{genres?.join(', ')}</p>
        </div>
      </Link>

      {children}
    </li>
  );
}

ResultListItemComponent.propTypes = {
  ...SharedMovieType,
};

export default ResultListItemComponent;
