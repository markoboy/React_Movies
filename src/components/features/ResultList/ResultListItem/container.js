/* eslint-disable camelcase */
import { MOVIE_FALLBACK_IMAGE } from '@constants/Generic';
import { SharedMovieType } from '@constants/MovieTypes';
import useFallbackImage from '@hooks/UseFallbackImage';
import React from 'react';
import ResultListItemComponent from './component';

function ResultListItemContainer({
  id,
  poster_path,
  title,
  release_date,
  genres,
  children,
}) {
  const { imageSrc, setFallbackImageSrc } = useFallbackImage(
    poster_path,
    MOVIE_FALLBACK_IMAGE
  );

  return (
    <ResultListItemComponent
      id={id}
      title={title}
      release_date={release_date}
      genres={genres}
      poster_path={imageSrc}
      onError={setFallbackImageSrc}
    >
      {children}
    </ResultListItemComponent>
  );
}

ResultListItemContainer.propTypes = SharedMovieType;

export default ResultListItemContainer;
