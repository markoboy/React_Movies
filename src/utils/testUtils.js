/* eslint-disable import/no-extraneous-dependencies */
import defaultStore from '@store/index';
import { render } from '@testing-library/react';
import React from 'react';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';

export function renderWithStore(
  ui,
  { store = defaultStore, route = '/', ...renderOptions } = {}
) {
  window.history.pushState({}, 'Test page', route);

  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }

  return render(ui, {
    wrapper: Wrapper,
    ...renderOptions,
  });
}

export function renderWithRouter(ui, { route = '/' } = {}) {
  window.history.pushState({}, 'Test page', route);

  return render(ui, { wrapper: BrowserRouter });
}
