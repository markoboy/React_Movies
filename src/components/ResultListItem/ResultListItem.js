/* eslint-disable jsx-a11y/anchor-is-valid */
import { MOVIE_FALLBACK_IMAGE } from '@constants/Generic';
import { MovieListItemType } from '@constants/MovieTypes';
import useFallbackImage from '@hooks/UseFallbackImage';
import React from 'react';
import './ResultListItem.scss';

function ResultListItem({
  image,
  title,
  releaseDate,
  genres,
  onClick,
  fallbackImage,
  children,
}) {
  const { imageSrc, setFallbackImageSrc } = useFallbackImage(image, fallbackImage);

  return (
    <li className="result-list__item">
      <a className="result-item-container" href="#" onClick={onClick}>
        <img
          src={imageSrc}
          alt={title}
          onError={setFallbackImageSrc}
        />

        <div className="result-item__body flex flex--wrap">
          <h4 className="result-item__title">{title}</h4>
          <p className="result-item__release-date">
            {releaseDate.getFullYear()}
          </p>
          <p className="result-item__genre">{genres.join(', ')}</p>
        </div>
      </a>

      {children}
    </li>
  );
}

ResultListItem.defaultProps = {
  fallbackImage: MOVIE_FALLBACK_IMAGE,
};

ResultListItem.propTypes = MovieListItemType;

export default ResultListItem;
