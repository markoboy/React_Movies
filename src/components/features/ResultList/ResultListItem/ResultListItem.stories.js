import React from 'react';

import withStoryBackgroundWrapper from '@components/hocs/WithStoryBackgroundWrapper';
import PosterImg from '@resources/avengers.jpg';
import ResultListItem from './index';

export default {
  component: ResultListItem,
  title: 'Feature/ResultList/ResultListItem',
  decorators: [
    withStoryBackgroundWrapper,
    (Story) => (
      <div style={{ maxWidth: '330px' }}>
        <Story />
      </div>
    ),
  ],
};

const Template = (args) => <ResultListItem {...args} />;

export const Default = Template.bind({});
Default.args = {
  id: 1,
  poster_path: PosterImg,
  title: 'Avengers',
  release_date: new Date(),
  genres: ['Action'],
};
