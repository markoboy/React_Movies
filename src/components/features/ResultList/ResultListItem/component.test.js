import React from 'react';

import ResultListItem from './component';
import { renderWithRouter } from '@utils/testUtils';

describe('ResultListItem Component', () => {
  it('renders ResultListItem Component with provided props', () => {
    const onError = jest.fn();

    const { container } = renderWithRouter(
      <ResultListItem
        id={1}
        poster_path="https://a-path.to/an/image.jpg"
        title="The best movie of all times"
        release_date={new Date()}
        genres={['Documentary']}
        onError={onError}
      />
    );

    expect(container).toMatchSnapshot();
  });
});
