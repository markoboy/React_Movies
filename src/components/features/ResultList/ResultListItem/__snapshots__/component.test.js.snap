// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultListItem Component renders ResultListItem Component with provided props 1`] = `
<div>
  <li
    class="result-list__item"
    data-testid="result-list-item"
  >
    <a
      class="result-item-container"
      href="/film/1"
    >
      <img
        alt="The best movie of all times"
        height="500"
        src="https://a-path.to/an/image.jpg"
        width="333"
      />
      <div
        class="result-item__body flex flex--wrap"
      >
        <h4
          class="result-item__title"
        >
          The best movie of all times
        </h4>
        <p
          class="result-item__release-date"
        >
          2021
        </p>
        <p
          class="result-item__genre"
        >
          Documentary
        </p>
      </div>
    </a>
  </li>
</div>
`;
